atc-router.go

```diff
@@ -113,3 +115,95 @@ func (r *Router) GetFields() ([]string, error) {
 
 	return flds, nil
 }
+
+const (
+	defaultNumFields    = 100
+	defaultMaxFieldSize = 100
+	errBufsize          = 500
+)
+
+var (
+	fieldsBuf = []C.uchar{}
+	fieldsLen = C.ulong(defaultMaxFieldSize * defaultNumFields)
+	fieldsNum = C.ulong(defaultNumFields)
+	errorBuf  = [errBufsize]C.uchar{}
+)
+
+type ValidationResult struct {
+	fields    []string
+	operators uint64
+	errorMsg  string
+}
+
+func ValidateExpression(s Schema, atc string) *ValidationResult {
+	atcC := unsafe.Pointer(C.CString(atc))
+	defer C.free(atcC)
+
+	operators := C.ulong(0)
+	errorBufLen := C.ulong(errBufsize)
+
+loop:
+	for {
+		if len(fieldsBuf) < int(fieldsLen) {
+			fieldsBuf = make([]C.uchar, fieldsLen)
+		}
+
+		switch C.expression_validate(
+			(*C.uchar)(atcC), s.s,
+			(*C.uchar)(&fieldsBuf[0]), &fieldsLen, &fieldsNum,
+			&operators,
+			&errorBuf[0], &errorBufLen) {
+		case 0:
+			break loop
+		case 1:
+			return &ValidationResult{
+				errorMsg: C.GoStringN(
+					(*C.char)(unsafe.Pointer(&errorBuf[0])),
+					C.int(errorBufLen),
+				),
+			}
+		case 2:
+			continue
+		}
+	}
+	return &ValidationResult{
+		fields:    splitByNulls(fieldsBuf, int(fieldsNum)),
+		operators: uint64(operators),
+	}
+}
+
+const (
+	BinaryOperatorFlags_EQUALS = 1 << iota
+	BinaryOperatorFlags_NOT_EQUALS
+	BinaryOperatorFlags_REGEX
+	BinaryOperatorFlags_PREFIX
+	BinaryOperatorFlags_POSTFIX
+	BinaryOperatorFlags_GREATER
+	BinaryOperatorFlags_GREATER_OR_EQUAL
+	BinaryOperatorFlags_LESS
+	BinaryOperatorFlags_LESS_OR_EQUAL
+	BinaryOperatorFlags_IN
+	BinaryOperatorFlags_NOT_IN
+	BinaryOperatorFlags_CONTAINS
+)
+
+func splitByNulls(b []C.uchar, maxN int) []string {
```

Comment:
I don't know whether it's a good idea or not but I was wondering what you think about using https://pkg.go.dev/bytes#Split instead of writing a custom split here. We'll need to cast C.uchar to byte array and that's equivalent to each other since they're both 8 bits but I'm not sure if there might be other issues.

Essentially, I was thinking something like: https://go.dev/play/p/H1i3X8BSJr7 (doesn't run in playground because cgo needs to be enabled but you can run it in a local main.go file)

Reply:
that was my first attempt, but the inconsistent use of `char`, `uchar`, `ichar`, `byte` makes the casts even more horrible than this solution.
in any case, the resultant slices would still have to pass through `C.GoStringN`. otherwise the slices still point to the C space.

Reply:
> but the inconsistent use of char, uchar, ichar, byte makes the casts even more horrible than this solution

I don't think I understand it correctly but I think you're saying these types are inconsistently used in the atc-router lib. Can we use generics in this function then?

> in any case, the resultant slices would still have to pass through C.GoStringN

Okay, I wasn't aware of this. Thanks!

I will let you decide but I just fear we might introduce some bugs in implementations at lower level. Can we add tests for this utility?

Reply:
> I don't think I understand it correctly but I think you're saying these types are inconsistently used in the atc-router lib.

yes, the atc-router uses both `*const u8` (for input) and `*mut u8` (for output buffers) in this function, while previous ones `*const i8` or other combinations.  also: `C.GoStringN` takes a `*C.char`, but strings usually cast to/from `[]byte`.  They're all pointers to bytes internally, (except `[]byte`, which also carries the length) but some can be directly cast and some others have to pass through `unsafe.Pointer`.  it's a real mess.

> Can we use generics in this function then?

i'll see if that helps, but i'm not sure.  after all, it's used on a single case so keeping it specific should be best... i'll move things around and see if can be improved.

> Can we add tests for this utility?

sure!  i did put a test to the exported function, but the internal one is definitely a fragile thing.  some tests could absolutely reduce future problems.